stdlib/src/Luna/Std/Finalizers.hs

```diff
@@ -26,7 +26,16 @@ registerFinalizer ctx finalizer = do
     return uuid
 
 cancelFinalizer :: FinalizersCtx -> UUID -> IO ()
-cancelFinalizer ctx uuid = modifyMVar_ (unwrap ctx) $ return . (wrapped . at uuid .~ Nothing)
+cancelFinalizer ctx uuid = do
+    -- [MM]: tc seems to deadlock if this function uses modifyMVar_ here
+    -- I suspect that is because this function can be run under another
+    -- modifyMVar_, from finalize below
```

Comment:
Sooo, if that would result in a deadlock, the finalizer will quietly not deregister? That's not really a solution :(

Reply:
A better solution would be to not use modifyMVar_ inside finalize, but run the IO actions with the MVar unlocked instead. I.e. get the contents of mvar, immediately put an empty finalizers map there and only then run the actual finalizers. Or the other way around – get the current contents, run the finalizers, put an empty map. Either solution should work. Even more predictable (though arguably slower) behavior would be to run and deregister the finalizers one-by-one in `finalize`, getting a fresh state after each finalizer is run (so that we do not run a finalizer twice).

Reply:
It will be dropped. This is an issue if during running finalize, one of our registered actions call cancelFinalizer - finalize emptied the MVar, and now cancelFinalizer waits on takeMVar.

I think the actual call stack is:
1. Backend calls finalize finalizersCtx disguised as cleanUp as an async exception handler
2. finalize takes our finalizers MVar
3. finalize calls uninterruptibleCancel on all of running Asyncs
4. Each Async receive an async exception
5. Handler for async exception calls cancelFinalizer that tries to take the MVar
6. Deadlock

Reply:
Yeah, I get the current scenario. And the current implementation (before your commit) is clearly broken. What I'm saying is: being able to depend on `deregisterFinalizer` to actually deregister the finalizer is important – the guarantee that any finalizer will be called at most once is important. With the proposed solution, if you try to kill some thread and at the same time free a large chunk of C memory, you may not deregister the freeing finalizer and call it again then. So we should go for a solution that provides basic guarantess, particularly given that it is possible.